### PR TITLE
Ticket 6418 - Check for Device Screens

### PR DESCRIPTION
**Pull request by aaron-long**
Added the ability to check for which Device Screens are present on a given instrument.

A test, execute the `get_ioc_usage.bat --device_screens`, which will output the names of the Device Screens present on each instrument.

- Some instruments have no devices screens and as such the returned list will be empty
- Currently we only check the `devices.xml` but I have created another ticket for checking the Synoptic view entries
- Due to the way device screen configs are names, there is no easy way to compare them against an associated driver (i.e., large variance in names "Aldn100 Pump" versus "ALDN1000_01").

### CODE REVIEW

File: get_ioc_usage.py [hunk 1]
```diff
@@ -78,6 +99,8 @@ def main():
                              "given instruments. If not defined, tests will be run on all instruments.")
     parser.add_argument("--ioc", type=str, default=None, help="If specified show instruments with IOC starting with"
                                                               "this string, otherwise show all iocs on the instruments")
+    parser.add_argument("--device_screens", help="Show if an instrument is missing an IOC",
```
Comment (rerpha) [inline get_ioc_usage.py:102]: I think the help string could be more descriptive and say something like "Show if an instrument has device screens that reference a missing IOC" or even just "print IOCs in device screens for instruments"

File: get_ioc_usage.py [hunk 2]
```diff
@@ -32,7 +33,17 @@ def calc_iocs_on_intruments(instruments):
...
 
-    return instrument_configs
+        device_utils = DeviceUtils(Settings.config_repo_path)
+        try:
+            device_screens = device_utils.get_device_screens(device_utils.get_device_screens_from_xml())
+        except Exception:
+            # Instrument does not have device screen directory
+            device_screens = []
+
+        instrument_name = instrument["name"]
```
Comment (rerpha) [inline get_ioc_usage.py:43]: I think `instrument_name` already exists as it's created above - might be able to remove this line?

File: get_ioc_usage.py [hunk 3]
```diff
@@ -52,12 +63,22 @@ def print_instruments_with_ioc(instrument_configs, ioc_name):
     :param instrument_configs: instrument ioc config dictionary
     :param ioc_name: name of the ioc
     """
-    print(f"Instruments containing IOCs starting with {ioc_name}")
+    print("Instruments containing IOCs starting with {}".format(ioc_name))
```
Comment (rerpha) [inline get_ioc_usage.py:66]: I think we should be able to use f-strings now as we're only using python 3 - it doesn't really matter but if you think it's more readable with or without them I think it's OK to use either.

File: util/configurations.py [hunk 4]
```diff
@@ -144,6 +145,33 @@ def _get_pv_name_without_field(pv_name):
...
+        Returns: the XML as a string
+        """
+        path = os.path.join(self.get_devices_directory(), "screens.xml")
+        with open(path) as xml_file:
+            return xml_file.read()
+
+    def get_device_screens(self, xml):
+        """
+        Args:
+            xml: XML inout to parse
```
Comment (rerpha) [inline util/configurations.py:163]:
```suggestion
            xml: XML input to parse
```

File: util/configurations.py [hunk 5]
```diff
@@ -144,6 +145,33 @@ def _get_pv_name_without_field(pv_name):
...
+        Returns:
+            device_screens (list): List of device screens
+        """
+        root = ET.fromstring(xml)
+
+        device_screens = []
+        for device in root:
+            device_screens.append(device.find(f"{self.DEVICES_XML_SCHEMA}key").text)
+
+        return set(device_screens)
```
Comment (rerpha) [inline util/configurations.py:173]: up to you whether you think this is nicer, but you can do this with set comprehension

```suggestion
        return {device.find(f"{self.DEVICES_XML_SCHEMA}key").text for device in root}
```

**Review by rerpha — COMMENTED** [review]
Given that we can't directly map IOCs to their device screens I think this is the best approach. I have made a couple of comments, but they're probably not big enough for me to stick the ticket in rework, and you might disagree with some anyway

**Comment by aaron-long** [comment]
Updated with the good changes you've suggested. The set comprehension is nice! I've omitted the format change as I think it's less readable.

**Review by rerpha — APPROVED** [review]

**Pull request merged**